src/libjson/jsonprint.c

```diff
@@ -0,0 +1,295 @@
+/*
+**==============================================================================
+**
+** Copyright (c) Microsoft Corporation
+**
+** All rights reserved.
+**
+** MIT License
+**
+** Permission is hereby granted, free of charge, to any person obtaining a copy
+** of this software and associated documentation files (the ""Software""), to
+** deal in the Software without restriction, including without limitation the
+** rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+** sell copies of the Software, and to permit persons to whom the Software is
+** furnished to do so, subject to the following conditions: The above copyright
+** notice and this permission notice shall be included in all copies or
+** substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+** OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+** THE SOFTWARE.
+**
+**==============================================================================
+*/
+
+#include <stdio.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "json.h"
+#include "jsonprint.h"
+#include "raise.h"
+
+static void _Indent(FILE* os, size_t depth)
+{
+    size_t i;
+
+    for (i = 0; i < depth; i++)
+        fprintf(os, "  ");
```

Comment:
Cherry on top: if we made `fprintf` configurable, I think we could hook it up to SGX-LKL's logging facilities (which will change soon too). Would be nice for error messages, etc.

Reply:
Nice suggestion. Done. This also eliminated the need for a second **libjsonprint.a** library. Now you can pass in a **write** callback function with a callback argument. This really cleaned things up but required implement a function to convert double-precision to string.